### PR TITLE
Streaming headers

### LINKED ISSUE #1627: thoughts on deserializing into an iterator (and not a simple vec)
Currently we read a full vec of elements during deserialization.

https://github.com/mimblewimble/grin/blob/4d2cbe6596462e609a528182e57a965ca4d5aa8d/core/src/ser.rs#L452-L461

We use this when handling a `Headers` msg during header sync.

I _think_ this could potentially be reworked (with a bit of work) to avoid reading the full vec in, returning an `iterator` instead.

Presumably we'd need to keep a ref to the underlying `reader` in the iterator.

But this would give us a ton of flexibility around exactly how/when we read the elements of the vec.

So we'd be able to do things like (code might look something like this) - 

```
headers = msg.body_iter()?;
let chunks = headers.chunks(8);
while let Some(header_chunk) = chunks.next() {
    adapter.headers_received(header_chunk);
}
```

This would allow us to avoid the custom headers deserialization logic where we estimate header size in bytes etc.
We'd be able to effectively deserialize them one at a time or several at a time then hand them off for further processing.

The actual header deserialization would be hidden behind the `body_iter()` (or something similar).

Thoughts @garyyu ?

### PR DESCRIPTION
**Pull request by antiochp**
We can now stream headers without estimating how many bytes we need to read and in a generalized way we can potentially leverage for other protocol messages.

* `msg.streaming_read()` reads a single `T` off the stream (`BlockHeader` for example).
* `StreamingReader` reads directly from a tcp stream
  * deserialize individual elements (via the underlying `BinReader`) on the fly
  * track total bytes read as we go (verify correct number of bytes read at the end)

Turns out deserializing off the tcp stream and into an iterator gets pretty complicated in rust because of the various lifetimes involved (conn, reader etc.)
But in protocol.rs it is actually easy enough to apply custom logic in the `Headers` handler allowing us to iterate over the headers _as the bytes arrive_.

This opens up some flexibility for how we handle full blocks - we can deserialize the header first and process it via the pipeline, then if everything is valid deserialize the rest of the full block and process it.
(Future work in a future PR).

We also have some flexibility now in terms of deserializing large vecs of inputs/outputs/kernels and potentially processing these in a similar (streaming) way.

Resolves #1627.

### CODE REVIEW

**Comment by antiochp** [comment]
```
20181115 14:28:34.404 DEBUG grin_servers::common::types - sync_state: sync_status: Initial -> AwaitingPeers(true)
20181115 14:28:41.429 DEBUG grin_servers::common::types - sync_state: sync_status: AwaitingPeers(true) -> HeaderSync { current_height: 0, highest_height: 40693 }
20181115 14:28:41.868 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 0, highest_height: 40693 } -> HeaderSync { current_height: 511, highest_height: 40693 }
20181115 14:28:42.192 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 511, highest_height: 40693 } -> HeaderSync { current_height: 1022, highest_height: 40693 }
20181115 14:28:42.804 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 1022, highest_height: 40693 } -> HeaderSync { current_height: 1533, highest_height: 40693 }
20181115 14:28:43.271 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 1533, highest_height: 40693 } -> HeaderSync { current_height: 2044, highest_height: 40693 }
20181115 14:28:43.614 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 2044, highest_height: 40693 } -> HeaderSync { current_height: 2555, highest_height: 40693 }
20181115 14:28:43.976 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 2555, highest_height: 40693 } -> HeaderSync { current_height: 3066, highest_height: 40693 }
20181115 14:28:44.625 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 3066, highest_height: 40693 } -> HeaderSync { current_height: 3577, highest_height: 40693 }
20181115 14:28:45.015 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 3577, highest_height: 40693 } -> HeaderSync { current_height: 4088, highest_height: 40693 }
20181115 14:28:45.412 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 4088, highest_height: 40693 } -> HeaderSync { current_height: 4599, highest_height: 40693 }
20181115 14:28:46.227 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 4599, highest_height: 40693 } -> HeaderSync { current_height: 5110, highest_height: 40693 }
20181115 14:28:47.543 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 5110, highest_height: 40693 } -> HeaderSync { current_height: 5621, highest_height: 40693 }
20181115 14:28:48.559 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 5621, highest_height: 40693 } -> HeaderSync { current_height: 6132, highest_height: 40693 }
20181115 14:28:50.046 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 6132, highest_height: 40693 } -> HeaderSync { current_height: 6643, highest_height: 40693 }
20181115 14:28:51.063 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 6643, highest_height: 40693 } -> HeaderSync { current_height: 7154, highest_height: 40693 }
20181115 14:28:52.022 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 7154, highest_height: 40693 } -> HeaderSync { current_height: 7665, highest_height: 40693 }
20181115 14:28:53.316 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 7665, highest_height: 40693 } -> HeaderSync { current_height: 8176, highest_height: 40693 }
20181115 14:28:54.865 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 8176, highest_height: 40693 } -> HeaderSync { current_height: 8687, highest_height: 40693 }
20181115 14:28:55.483 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 8687, highest_height: 40693 } -> HeaderSync { current_height: 9198, highest_height: 40693 }
20181115 14:28:56.407 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 9198, highest_height: 40693 } -> HeaderSync { current_height: 9709, highest_height: 40693 }
20181115 14:28:57.266 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 9709, highest_height: 40693 } -> HeaderSync { current_height: 10220, highest_height: 40693 }
20181115 14:28:57.727 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 10220, highest_height: 40693 } -> HeaderSync { current_height: 10731, highest_height: 40693 }
20181115 14:28:58.296 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 10731, highest_height: 40693 } -> HeaderSync { current_height: 11242, highest_height: 40693 }
20181115 14:28:58.889 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 11242, highest_height: 40693 } -> HeaderSync { current_height: 11753, highest_height: 40693 }
20181115 14:28:59.673 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 11753, highest_height: 40693 } -> HeaderSync { current_height: 12264, highest_height: 40693 }
20181115 14:29:00.664 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 12264, highest_height: 40693 } -> HeaderSync { current_height: 12775, highest_height: 40693 }
20181115 14:29:01.239 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 12775, highest_height: 40693 } -> HeaderSync { current_height: 13286, highest_height: 40693 }
20181115 14:29:01.699 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 13286, highest_height: 40693 } -> HeaderSync { current_height: 13797, highest_height: 40693 }
20181115 14:29:02.595 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 13797, highest_height: 40693 } -> HeaderSync { current_height: 14308, highest_height: 40693 }
20181115 14:29:03.053 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 14308, highest_height: 40693 } -> HeaderSync { current_height: 14819, highest_height: 40693 }
20181115 14:29:03.494 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 14819, highest_height: 40693 } -> HeaderSync { current_height: 15330, highest_height: 40693 }
20181115 14:29:04.039 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 15330, highest_height: 40693 } -> HeaderSync { current_height: 15841, highest_height: 40693 }
20181115 14:29:04.542 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 15841, highest_height: 40693 } -> HeaderSync { current_height: 16352, highest_height: 40693 }
20181115 14:29:05.568 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 16352, highest_height: 40693 } -> HeaderSync { current_height: 16863, highest_height: 40693 }
20181115 14:29:06.612 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 16863, highest_height: 40693 } -> HeaderSync { current_height: 17374, highest_height: 40693 }
20181115 14:29:07.607 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 17374, highest_height: 40693 } -> HeaderSync { current_height: 17885, highest_height: 40693 }
20181115 14:29:08.988 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 17885, highest_height: 40693 } -> HeaderSync { current_height: 18396, highest_height: 40693 }
20181115 14:29:09.777 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 18396, highest_height: 40693 } -> HeaderSync { current_height: 18907, highest_height: 40693 }
20181115 14:29:10.461 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 18907, highest_height: 40693 } -> HeaderSync { current_height: 19418, highest_height: 40693 }
20181115 14:29:11.914 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 19418, highest_height: 40693 } -> HeaderSync { current_height: 19929, highest_height: 40693 }
20181115 14:29:12.889 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 19929, highest_height: 40693 } -> HeaderSync { current_height: 20440, highest_height: 40693 }
20181115 14:29:14.034 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 20440, highest_height: 40693 } -> HeaderSync { current_height: 20951, highest_height: 40693 }
20181115 14:29:15.103 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 20951, highest_height: 40693 } -> HeaderSync { current_height: 21462, highest_height: 40693 }
20181115 14:29:16.103 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 21462, highest_height: 40693 } -> HeaderSync { current_height: 21973, highest_height: 40693 }
20181115 14:29:16.878 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 21973, highest_height: 40693 } -> HeaderSync { current_height: 22484, highest_height: 40693 }
20181115 14:29:17.761 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 22484, highest_height: 40693 } -> HeaderSync { current_height: 22995, highest_height: 40693 }
20181115 14:29:18.316 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 22995, highest_height: 40693 } -> HeaderSync { current_height: 23506, highest_height: 40693 }
20181115 14:29:18.906 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 23506, highest_height: 40693 } -> HeaderSync { current_height: 24017, highest_height: 40693 }
20181115 14:29:19.578 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 24017, highest_height: 40693 } -> HeaderSync { current_height: 24528, highest_height: 40693 }
20181115 14:29:20.382 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 24528, highest_height: 40693 } -> HeaderSync { current_height: 25039, highest_height: 40693 }
20181115 14:29:20.935 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 25039, highest_height: 40693 } -> HeaderSync { current_height: 25550, highest_height: 40693 }
20181115 14:29:21.653 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 25550, highest_height: 40693 } -> HeaderSync { current_height: 26061, highest_height: 40693 }
20181115 14:29:22.454 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 26061, highest_height: 40693 } -> HeaderSync { current_height: 26572, highest_height: 40693 }
20181115 14:29:23.067 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 26572, highest_height: 40693 } -> HeaderSync { current_height: 27083, highest_height: 40693 }
20181115 14:29:23.755 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 27083, highest_height: 40693 } -> HeaderSync { current_height: 27594, highest_height: 40693 }
20181115 14:29:24.367 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 27594, highest_height: 40693 } -> HeaderSync { current_height: 28105, highest_height: 40693 }
20181115 14:29:25.029 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 28105, highest_height: 40693 } -> HeaderSync { current_height: 28616, highest_height: 40693 }
20181115 14:29:25.746 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 28616, highest_height: 40693 } -> HeaderSync { current_height: 29127, highest_height: 40693 }
20181115 14:29:27.128 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 29127, highest_height: 40693 } -> HeaderSync { current_height: 29638, highest_height: 40693 }
20181115 14:29:28.448 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 29638, highest_height: 40693 } -> HeaderSync { current_height: 30149, highest_height: 40693 }
20181115 14:29:29.555 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 30149, highest_height: 40693 } -> HeaderSync { current_height: 30660, highest_height: 40693 }
20181115 14:29:30.633 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 30660, highest_height: 40693 } -> HeaderSync { current_height: 31171, highest_height: 40693 }
20181115 14:29:31.391 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 31171, highest_height: 40693 } -> HeaderSync { current_height: 31682, highest_height: 40693 }
20181115 14:29:32.006 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 31682, highest_height: 40693 } -> HeaderSync { current_height: 32193, highest_height: 40693 }
20181115 14:29:32.605 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 32193, highest_height: 40693 } -> HeaderSync { current_height: 32704, highest_height: 40693 }
20181115 14:29:33.324 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 32704, highest_height: 40693 } -> HeaderSync { current_height: 33215, highest_height: 40693 }
20181115 14:29:34.159 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 33215, highest_height: 40693 } -> HeaderSync { current_height: 33726, highest_height: 40693 }
20181115 14:29:34.875 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 33726, highest_height: 40693 } -> HeaderSync { current_height: 34237, highest_height: 40693 }
20181115 14:29:35.867 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 34237, highest_height: 40693 } -> HeaderSync { current_height: 34748, highest_height: 40693 }
20181115 14:29:36.496 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 34748, highest_height: 40693 } -> HeaderSync { current_height: 35259, highest_height: 40693 }
20181115 14:29:37.148 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 35259, highest_height: 40693 } -> HeaderSync { current_height: 35770, highest_height: 40693 }
20181115 14:29:37.919 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 35770, highest_height: 40693 } -> HeaderSync { current_height: 36281, highest_height: 40693 }
20181115 14:29:38.582 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 36281, highest_height: 40693 } -> HeaderSync { current_height: 36792, highest_height: 40693 }
20181115 14:29:39.486 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 36792, highest_height: 40693 } -> HeaderSync { current_height: 37303, highest_height: 40693 }
20181115 14:29:40.141 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 37303, highest_height: 40693 } -> HeaderSync { current_height: 37814, highest_height: 40693 }
20181115 14:29:40.812 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 37814, highest_height: 40693 } -> HeaderSync { current_height: 38325, highest_height: 40693 }
20181115 14:29:41.776 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 38325, highest_height: 40693 } -> HeaderSync { current_height: 38836, highest_height: 40693 }
20181115 14:29:42.489 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 38836, highest_height: 40693 } -> HeaderSync { current_height: 39347, highest_height: 40693 }
20181115 14:29:43.176 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 39347, highest_height: 40693 } -> HeaderSync { current_height: 39858, highest_height: 40693 }
20181115 14:29:44.005 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 39858, highest_height: 40693 } -> HeaderSync { current_height: 40369, highest_height: 40693 }
20181115 14:29:45.511 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 40369, highest_height: 40693 } -> HeaderSync { current_height: 40694, highest_height: 40694 }
20181115 14:29:45.511 DEBUG grin_servers::common::types - sync_state: sync_status: HeaderSync { current_height: 40694, highest_height: 40694 } -> TxHashsetDownload { start_time: 2018-11-15T14:29:45.511774Z, downloaded_size: 0, total_size: 0 }
```

File: p2p/src/msg.rs [hunk 1]
```diff
@@ -213,6 +213,92 @@ pub fn read_header(conn: &mut TcpStream, msg_type: Option<Type>) -> Result<MsgHe
 	Ok(header)
 }
 
+/// A reader that reads straight off a stream.
```
Comment (ignopeverell) [inline p2p/src/msg.rs]: Considered putting that in core `ser.rs`? Seems that it could be reused in other places over time.

Comment (ignopeverell) [reply p2p/src/msg.rs]: Nevermind, hadn't realized it relied on our `read_exact` definition.

Comment (antiochp) [reply p2p/src/msg.rs]: Done.
It has a dependency on our custom `read_exact()` so moved that out of `p2p` and into `util`. This now uses a generic `Read` and not a specific `TcpStream`.

**Pull request merged**